[SPARK-41396][SQL][PROTOBUF] OneOf field support and recursion checks

SandishKumarHN opened the pull request:
Oneof fields allow a message to contain one and only one of a defined set of field types, while recursive fields provide a way to define messages that can refer to themselves, allowing for the creation of complex and nested data structures.  with this change users will be able to use protobuf OneOf fields with spark-protobuf, making it a more complete and useful tool for processing protobuf data.

**Support for circularReferenceDepth:**
The `recursive.fields.max.depth` parameter can be specified in the from_protobuf options to control the maximum allowed recursion depth for a field. Setting `recursive.fields.max.depth` to 0 drops all-recursive fields, setting it to 1 allows it to be recursed once, and setting it to 2 allows it to be recursed twice. Attempting to set the `recursive.fields.max.depth` to a value greater than 10 is not allowed. If the `recursive.fields.max.depth` is not specified, it will default to -1; recursive fields are not permitted. if a protobuf record has more depth for recursive fields than the allowed value, it will be truncated and some fields may be discarded. This check is based on the fully qualified field type.
SQL Schema for the protobuf message
 ```
message Person { 
     string name = 1;
     Person bff = 2
}
```
will vary based on the value of `recursive.fields.max.depth`.
```
0: struct<name: string, bff: null>
1: struct<name string, bff: <name: string, bff: null>>
2: struct<name string, bff: <name: string, bff: struct<name: string, bff: null>>> ...
```

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
- Add support for protobuf oneof field
- Stop recursion at the first level when a recursive field is encountered. (instead of throwing an error) 

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Stop recursion at the first level and handle nulltype in deserilization. 


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

NA
### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Added Unit tests for OneOf field support and recursion checks.
Tested full support for nested OneOf fields and message types using real data from Kafka on a real cluster

cc: @rangadi @mposdev21

SandishKumarHN committed:
OneOf field support and recursion checks

SandishKumarHN commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,9 +92,13 @@ object SchemaConverters {
             nullable = false))
       case MESSAGE =>
+        // Stop recursion at the first level when a recursive field is encountered.
+        // TODO: The user should be given the option to set the recursion level to 1, 2, or 3
```
@rangadi Instead of limiting the recursion to only one level, the user should be able to choose a recursion level of 1, 2, or 3. Going beyond 3 levels of recursion should not be allowed. any thoughts? 

`spark.protobuf.recursion.level`

rangadi replied:
Yeah, I think it is useful. Users may not be able to remove recursive references, but might be willing to limit recursion.
I think the default should be an error with a clear message about how users can set configuration. 
Also, I don't think it should be spark config, but rather an `option` passed in.

rangadi replied:
Are you planning to add selectable recursion depth here or in a follow up?

SandishKumarHN replied:
@rangadi planning to add the selectable recursion depth in this PR.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala:
```diff
@@ -157,6 +157,8 @@ private[sql] class ProtobufDeserializer(
 
       case (null, NullType) => (updater, ordinal, _) => updater.setNullAt(ordinal)
 
+      case (MESSAGE, NullType) => (updater, ordinal, _) => updater.setNullAt(ordinal)
```
What is this for? For handling limited recursion?

SandishKumarHN replied:
yes, correct.

rangadi replied:
Could you add a comment about we might be dropping data here? It will not be easy to see for a future reader. 
We could have an option to error our if the actual data has more recursion than the configure.

rangadi commented on connector/protobuf/src/test/resources/protobuf/functions_suite.proto:
```diff
@@ -170,4 +170,41 @@ message timeStampMsg {
\ No newline at end of file
+}
+
+message OneOfEvent {
```
Are you testing more OneOf and recusion in the same message? Could you split them into separate messages?

SandishKumarHN replied:
@rangadi I see a lot of use cases for the "payload" Oneof the field and recursive fields in it. So I thought combining Oneof with recursion would be a good test. will separate

rangadi replied:
Combined one is fine, we could keep it. Better to have a simpler separate tests as well.

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +721,45 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
       parameters = Map("descFilePath" -> testFileDescriptor))
   }
+
+  test("Unit tests for OneOf field support and recursion checks") {
```
Lets separate these two into separate tests with separate protobuf message.

rangadi commented on connector/protobuf/src/test/resources/protobuf/functions_suite.proto:
```diff
@@ -170,4 +170,41 @@ message timeStampMsg {
+
+message OneOfEvent {
+  string key = 1;
+  oneof payload {
```
How do one-of fields look like in spark schema? Could you give an example? I could not see the schema in the unit tests.

SandishKumarHN replied:
@rangadi the "Oneof" field is of message type, Oneof will be converted to a struct type.

SandishKumarHN committed:
review changes, recursionDepth option, unit tests

SandishKumarHN commented:
https://github.com/apache/spark/pull/38922#discussion_r1041470191

@rangadi made the below changes.
- Added selectable recursion depth option to from_protobuf.
- Added two unit tests for Oneof type, a simple one for Oneof field, and a complex Onefield with recursionDepth=2.
- Existing unit tests should cover `foundRecursionInProtobufSchema` if recursionDepth is not set and a recursion field is discovered.

rangadi commented:
> Added selectable recursion depth option to from_protobuf.

Do we need to this for 'to_protobuf()' too? What would happen in that case?

SandishKumarHN commented:
> > Added selectable recursion depth option to from_protobuf.
> 
> Do we need to this for 'to_protobuf()' too? What would happen in that case?
@rangadi 
The source dataframe struct field should match the protobuf recursion message for "to protobuf." It will convert until the recursion level is matched. like struct within a struct to recursive message. This is true even for existing code.

rangadi commented:
> The source dataframe struct field should match the protobuf recursion message for "to protobuf." It will convert until the recursion level is matched. like struct within a struct to recursive message. This is true even for existing code.

Interesting. So we would make that null after some depth. Could you add test for this?

SandishKumarHN commented:
> > The source dataframe struct field should match the protobuf recursion message for "to protobuf." It will convert until the recursion level is matched. like struct within a struct to recursive message. This is true even for existing code.
> 
> Interesting. So we would make that null after some depth. Could you add test for this?

@rangadi will add a test for the above case. 
A Spark dataframe with complex nested structures should typically be converted to a protobuf message. It is the user's responsibility to specify right .proto(.desc) file that corresponds to the source dataframe.

rangadi commented:
>  file that corresponds to the source dataframe.

They might have used from_protobuf() to get that schema, which supports recursive fields. They should be able to do to_protobuf() with the same protobuf definition.

SandishKumarHN commented:
> >  file that corresponds to the source dataframe.
> 
> 
> 
> They might have used from_protobuf() to get that schema, which supports recursive fields. They should be able to do to_protobuf() with the same protobuf definition.

This case is already been covered in the unit tests. will add a unit test for direct struct to protobuf conversion.

SandishKumarHN committed:
review changes: adding unit tests for struct to protobuf

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-41396

AmplabJenkins commented:
Can one of the admins verify this patch?

baganokodo2022 commented:
Hi @SandishKumarHN,

For the `recursionDepth` option, could we consider naming it as `CircularReferenceTolerance` or `CircularReferenceDepth` for clarity?
For instance, -1 (default value) will error out on any circular reference, 0 drops any circular reference field, 1 allows the same field to be entered twice, and on.
 
Besides, can we also support a "CircularReferenceType" option with a enum value of `[FIELD_NAME, FIELD_TYPE]`. The reason is because navigation can go very deep before the same **fully-qualified** `FIELD_NAME` is encountered again. While `FIELD_TYPE` stops recursive navigation much faster. We could make `FIELD_NAME` the default option. In my test cases, with `FIELD_TYPE`, a circular reference can repeat 3 times before the executor hit OOM, while `FIELD_NAME` hit OOM when `CircularReferenceTolerance` is set to 1.

Please let me know your thoughts.

cc @rangadi 

Thank you

Xinyu Liu

SandishKumarHN committed:
circularReferenceType && circularReferenceDepth

SandishKumarHN commented:
@baganokodo2022 Circler type(specially MESSAGE) occurs frequently in a single message. The user won't be able to distinguish it and fix it (imagine which field should user keep or remove). because each type will have a unique field name that is valid. The user can verify and fix the circular reference in the full field name scenario. 

anyways, I have made the initial change as per your idea. have a look at it.

cc: @rangadi

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,14 +109,38 @@ object SchemaConverters {
-        if (existingRecordNames.contains(fd.getFullName)) {
-          throw QueryCompilationErrors.foundRecursionInProtobufSchema(fd.toString())
+        // User can set circularReferenceDepth of 0 or 1 or 2.
+        // Going beyond 3 levels of recursion is not allowed.
```
Could you add a justification for this?

SandishKumarHN replied:
@rangadi The user can specify the maximum allowed recursion depth for a field by setting the circularReferenceDepth property to 0, 1, or 2. Setting the circularReferenceDepth to 0 allows the field to be recursed once, setting it to 1 allows it to be recursed twice, and setting it to 2 allows it to be recursed thrice. Attempting to set the circularReferenceDepth to a value greater than 2 is not allowed. If the circularReferenceDepth is not specified, it will default to -1, which disables recursive fields.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -38,6 +38,12 @@ private[sql] class ProtobufOptions(
   val parseMode: ParseMode =
     parameters.get("mode").map(ParseMode.fromString).getOrElse(FailFastMode)
+
+  val circularReferenceType: String = parameters.getOrElse("circularReferenceType", "FIELD_NAME")
```
@SandishKumarHN @baganokodo2022 moving the discussion here (for threading).

> Besides, can we also support a "CircularReferenceType" option with a enum value of [FIELD_NAME, FIELD_TYPE]. The reason is because navigation can go very deep before the same fully-qualified FIELD_NAME is encountered again. While FIELD_TYPE stops recursive navigation much faster.  ...

I didn't quite follow the motivation here. Could you give a concrete examples for the two difference cases?

SandishKumarHN replied:
@rangadi we already know about field_name circusive check. using `fd.getFullName` we detect the recursion and throw and error.  another option is to detect recursion through field type. example below.

```
message A {
B b;
}

message B {
A c;
}
```
in the case of field_name recursive check it is ```A.B.C```  no recursion. 
in the case of field_type recursive check. it is ```MESSAGE.MESSAGE.MESSAGE``` recursion will be found and throw an error or drop the certain recursive depth. 
but it will also throw an error for the below case with the field_type check. since it will be ```MESSAGE.MESSAGE.MESSAGE.MESSAGE```

```
message A {
B b = 1;
}

message B {
D d = 1;
}

message D {
E e = 1;
}

message E {
int32 key = 1;
}
```
@baganokodo2022 argument is field_type base check will give users an option to drop recursion more quickly because with complex nested schema recursive field_name can be found at very deep. before hitting this we might see OOM. field_type base check finds the circle reference more quickly. 

@baganokodo2022 please correct me if I'm wrong.

rangadi replied:
> in the case of field_name recursive check it is A.B.C no recursion.

That example is clearly recursion. What is 'C' here. 

> but it will also throw an error for the below case with the field_type check. since it will be MESSAGE.MESSAGE.MESSAGE.MESSAGE

Why is this recursion?

rangadi replied:
Are our unit tests showing these cases?

SandishKumarHN replied:
I would have @baganokodo2022 give more details on the field type case. 

We have not yet added unit tests for the field-type case. would like to discuss this before adding unit tests.

rangadi replied:
@baganokodo2022 Could you translate that to actual protobuf to illustrate the problem? I still don't understand. 
If the redundat data in the warehouse is concern, customers can process with a smaller protobuf (say with unneeded fields removed).

baganokodo2022 replied:
@rangadi , the issue is that protobuf is designed for interactions between micro services. Protobuf schemas are long existing before we introduced Delta lake to the corporate. In many cases, we have to comprise for the needs of micro services, not the other way around. Service developers typically are not motivated or willing to take the risk to revise their data model for data analytics purposes. I will compile a proto schema shortly. cc @SandishKumarHN 

thanks

rangadi replied:
I see only one type of recursion, that is just just based on protobuf message. I.e. 
 message A ends up including itself either directly or indirectly through nesting. This is the 1st one in @SandishKumarHN's message above. The second one does not have recursion. 

Looking for examples of other types.

SandishKumarHN replied:
```
message A {
B b = 1;
}
message B {
A aa = 1;
D d = 2;
}

message D {
A aaa = 1;
E a =2;
}

message E {
int32 key = 1;
}
```
@rangadi just correcting the second example. in the case of field_name base recursive check. we would fail to detect the recursion for above because the thread would be ```A.B.A.aa.D.d.A.aaa.E``` set would have unique values [A, B.b, A.aa, A.aaa, E.a], but in the case of @baganokodo2022 proposal it would be ```A.B.A.D.A.E``` would takeout field_name so early detection for recursion.

SandishKumarHN replied:
@rangadi just replacing fd.getFullName with below would make sense. we don't need to have two different cases of checks. 
```
val recordName = fd.getFullName.substring(0, fd.getFullName().lastIndexOf("."))
```

rangadi replied:
> we would fail to detect the recursion for above because the thread would be

Why would we fail? Lets say this is doing `from_protobuf(col, 'message_A')`
`A aa = 1` at line 5 would be treated as recursion.
Why are field names relevant at all?

rangadi replied:
Does the code in master fail to detect this recursion? In that case it would lead to unbounded schema and fail.

rangadi replied:
>  thread would be `A.B.A.aa.D.d.A.aaa.E`

What is this thread?

rangadi replied:
Given this discussion, let's write down functionality and examples, before we implement so that we are all on the same page.

SandishKumarHN replied:
@rangadi fd.fullName is able to detect the recursive field with different field names. add a unit test. now I'm confused.  
```Fail for recursion field with different field names```

rangadi replied:
:) yeah, field names should not matter at all. 
We can do video chat to clarify all this.

SandishKumarHN replied:
@rangadi @baganokodo2022 thanks for the quick meet. meeting conclusion use descriptor type name and added unit tests with some complex schema. 
```
val recordName = fd.getMessageType.getFullName
```

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,14 +109,38 @@ object SchemaConverters {
+            (protobufOptions.circularReferenceDepth < 0 ||
+              protobufOptions.circularReferenceDepth >= 3)) {
+            throw QueryCompilationErrors.foundRecursionInProtobufSchema(fd.toString())
+          } else if (existingRecordTypes.contains(fd.getType.name()) &&
```
name or full name? 
also what keeps track of the recursion depth?

SandishKumarHN replied:
@rangadi we have two maps with incremental counters, one for field_name base check and one for field_type.

baganokodo2022 replied:
In my back-ported branch,
```
        val recordName = circularReferenceType match {
          case CircularReferenceTypes.FIELD_NAME =>
            fd.getFullName
          case CircularReferenceTypes.FIELD_TYPE =>
            fd.getFullName().substring(0, fd.getFullName().lastIndexOf(".")) 
        }
        
        if (circularReferenceTolerance < 0 && existingRecordNames(recordName) > 0) {
          // no tolerance on circular reference
          logError(s"circular reference in protobuf schema detected [no tolerance] - ${recordName}")
          throw new IllegalStateException(s"circular reference in protobuf schema detected [no tolerance] - ${recordName}")
        }

        if (existingRecordNames(recordName) > (circularReferenceTolerance max 0) ) {
          // stop navigation and drop the repetitive field
          logInfo(s"circular reference in protobuf schema detected [max tolerance breached] field dropped - ${recordName} = ${existingRecordNames(recordName)}")
          Some(NullType)
        } else {
          val newRecordNames: Map[String, Int] = existingRecordNames +  
            (recordName -> (1 + existingRecordNames(recordName)))
          Option(
            fd.getMessageType.getFields.asScala
              .flatMap(structFieldFor(_, newRecordNames, protobufOptions))
              .toSeq)
            .filter(_.nonEmpty)
            .map(StructType.apply)
        }```

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +693,178 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
+      assert(expectedFields.contains(f.getName))
+    })
+
+    val schema = StructType(Seq(StructField("sample",
```
Btw, using `val schema = DataType.fromJson("json string") is lot more readable. 
Optional we could update many of these in follow up PRs.

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +693,178 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
       parameters = Map("descFilePath" -> testFileDescriptor))
   }
+
+  test("Unit test for Protobuf OneOf field") {
```
Add a short description of the test at the top. It improves readability. What is this verifying? 

Remove "Unit test for", this is already a unit test :).

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -26,11 +26,11 @@ import com.google.protobuf.{ByteString, DynamicMessage}
 import org.apache.spark.sql.AnalysisException
 import org.apache.spark.sql.functions.{lit, struct}
-import org.apache.spark.sql.protobuf.protos.SimpleMessageProtos.SimpleMessageRepeated
+import org.apache.spark.sql.protobuf.protos.SimpleMessageProtos.{EventRecursiveA, EventRecursiveB, OneOfEvent, OneOfEventWithRecursion, SimpleMessageRepeated}
```
Are there tests for recursive fields?

SandishKumarHN replied:
@rangadi yes, 
`Handle recursive fields in Protobuf schema, C->D->Array(C)` and
`Handle recursive fields in Protobuf schema, A->B->A`

rangadi replied:
Could we move that to different tests?

SandishKumarHN replied:
@rangadi I didn't understand. these are already two different tests.

SandishKumarHN committed:
review changes, desc for unittests, jsonschema

baganokodo2022 reviewed (commented):
thank you for the PR

HeartSaVioR commented:
cc. @cloud-fan I guess there has been some demands on recursive schema already. Could you please help looking into the proposal and see whether it makes sense to you? Please add more ppl in the loop if you know others who would be interested.

HeartSaVioR commented:
I guess the demand on supporting recursive schema is not specific to protobuf, it also applies to Avro. If we construct a way how to project the recursive schema into Spark SQL's schema, we may want to apply this consistently across components.

The visibility of this PR is too limited, ppl interested in protobuf would only look into this. Instead of deciding such thing within this PR, it seems like going through discussion thread in dev@ for this is not a bad idea. What do you all think?

If you have a proposal, please write it down to the doc format e.g. google doc, with several examples. The description of the PR does not seem to be enough to understand what this PR (or some other) is proposing.

SandishKumarHN committed:
review changes, circularReferenceDepth changes

SandishKumarHN commented:
@baganokodo2022 instead of implementing a check for circular reference types in this PR, can we discuss this further and write a proposal for it before adding it to the next PR? We can share the proposal with the [dev@spark.org](mailto:dev@spark.org) mailing list for feedback and input.

@rangadi @HeartSaVioR In this PR, we will only implement a check for circular references through the full field name. Let me know if any further changes are needed.

rangadi commented:
@SandishKumarHN cold you keep the discussion in the code [review thread](https://github.com/apache/spark/pull/38922#discussion_r1043725083)? It is hard to piece together multiple messages. I think the it is fairly straight forward what recursion means. There seems to be some confusion about that. Let's discuss in the thread.

HeartSaVioR commented:
OK the recursion support isn't something the PR originally proposes. Got it. Then as long as it fits with what we support with Avro, it looks OK to me. cc. @gengliangwang

SandishKumarHN committed:
unit test for recursive field with different fieldnames

SandishKumarHN committed:
unit test for recursive field with different fieldnames

SandishKumarHN committed:
meeting conclusion, use decriptor fullname

cloud-fan commented:
Can we give an example mapping between recursive proto type and spark data type? There is no native recursion support in Spark's type system, so this is a bit counterintuitive.

SandishKumarHN commented:
> Can we give an example mapping between recursive proto type and spark data type? There is no native recursion support in Spark's type system, so this is a bit counterintuitive.

@cloud-fan sure, Since spark does not support circular references, so we are providing an option called circularReferenceDepth that allows users to specify at what point the recursion should stop. By default, this option will be disabled and an error will be thrown if a circular reference is detected.

```
message A {
B b = 1;
C c = 2
int32 key = 3;
}

message B {
A a = 1;
int32 key = 2;
}

message C {
A a = 1;
int32 key = 2; 
}
```

the above protobuf schema will be converted to a Spark struct where the struct contains a field of type struct within a struct. This allows the schema to support nested data structures.

default value -1: will throw an error if we find any recursive fields,

level 0 :
```
A = Struct(b = Struct(a = Struct(), key = 123), c = Struct(a = Struct(), key = 123), key = 123)
```
level 1:
```
A = Struct(b = Struct(a = Struct(b = Struct(), c = Struct(), key = 123), key = 123), c = Struct(a = Struct(b = Struct(), c = Struct(), key = 123), key = 123), key = 123)
```
level 2: 
```
A = Struct(b = Struct(a = Struct(b = Struct(a = Struct(), key = 123), c = Struct(a = Struct(), key = 123), key = 123), key = 123), c = Struct(a = Struct(b = Struct(a = Struct(), key = 123), c = Struct(a = Struct(), key = 123), key = 123), key = 123), key = 123)
```

mposdev21 commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala:
```diff
@@ -235,7 +237,7 @@ private[sql] class ProtobufDeserializer(
           updater.set(ordinal, row)
 
-      case (MESSAGE, ArrayType(st: StructType, containsNull)) =>
+      case (MESSAGE, ArrayType(st: DataType, containsNull)) =>
```
It's not clear as to why we need to make this change. It does not look like OneOf related. Could you clarify why this would be needed ? Does it cover any specific case ?

SandishKumarHN replied:
@mposdev21 Actually, we don't need this here at all; we can add it at the top along with other arraytype types.

SandishKumarHN committed:
review change code optimization

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -38,6 +38,12 @@ private[sql] class ProtobufOptions(
+  // Setting the circularReferenceDepth to 0 allows the field to be recursed once, setting
+  // it to 1 allows it to be recursed twice, and setting it to 2 allows it to be recursed
+  // thrice. circularReferenceDepth value greater than 2 is not allowed. If the not
+  // specified, it will default to -1, which disables recursive fields.
```
'-1' implies recursive fields are not allowed. 
("disables" does not clearly imply that it will be an error")

rangadi replied:
Also warn that if the the protobuf record has more depth for recursive fields than allowed here, it will be truncated to the allowed depth. The implies some fields are discarded from the record. 

Could you add a simple example in the comment showing resulting spark schema when this is set to '0' and '2'.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -38,6 +38,12 @@ private[sql] class ProtobufOptions(
+  // it to 1 allows it to be recursed twice, and setting it to 2 allows it to be recursed
+  // thrice. circularReferenceDepth value greater than 2 is not allowed. If the not
+  // specified, it will default to -1, which disables recursive fields.
+  val circularReferenceDepth: Int = parameters.getOrElse("circularReferenceDepth", "-1").toInt
```
Suggestion for renaming this option:  _"recursive.fields.max.depth"_ 

_circularReferenceDepth_ sounds very code variable type.

rangadi replied:
If we go with that, we could rename the variable as well to 'recursiveFieldMaxDepth' (but this is your choice).

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,14 +106,26 @@ object SchemaConverters {
+        } else if (existingRecordNames.contains(recordName) &&
+          existingRecordNames.getOrElse(recordName, 0)
+            > protobufOptions.circularReferenceDepth) {
+          return Some(StructField(fd.getName, NullType, nullable = false))
```
Why is nullable false?

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,14 +106,26 @@ object SchemaConverters {
+        // thrice. circularReferenceDepth value greater than 2 is not allowed. If the not
+        // specified, it will default to -1, which disables recursive fields.
+        val recordName = fd.getMessageType.getFullName
+        if (existingRecordNames.contains(recordName) &&
```
Better to remove 'return' statement. 
How about"

    val recursiveDepth = existingRecordNames.getOrElse(recordName, 0)
    if (recursiveDepth == 0 ||  // No recursion
      (protobufOptions.circularReferenceDepth >= 0 
        && recursiveDepth <=  (protobufOptions.circularReferenceDepth + 1)) // recursion is within allowed limit.
         val newRecordNames = existingRecordNames + (recordName -> recursiveDepth + 1)) 
          ... 
    } else if (protobufOptions.circularReferenceDepth >= 0) { 
         // Recursion is allowed we reached limit. Truncate.
      return Some(StructField(fd.getName, NullType, ...)      
    } else { // Recursion is not allowed
       throw ...
    }

rangadi replied:
Scratch the above suggestion. 
Instead you could add 'else' what you have and remove 'return'. That is simpler.

SandishKumarHN replied:
@rangadi thanks for the review, I have made all changes you suggested.

rangadi commented on connector/protobuf/src/test/resources/protobuf/functions_suite.proto:
```diff
@@ -170,4 +170,118 @@ message timeStampMsg {
+message Employee {
+  string firstName = 1;
+  string lastName = 2;
+  oneof role {
```
Do we need so many fields for 'OneOf'? How about just 2 or 3? It will simplify testing.

SandishKumarHN committed:
review changes, recursive.fields.max.depth, comments, nit

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -38,6 +38,14 @@ private[sql] class ProtobufOptions(
   val parseMode: ParseMode =
     parameters.get("mode").map(ParseMode.fromString).getOrElse(FailFastMode)
+
+  // Setting the `recursive.fields.max.depth` to 0 allows the field to be recurse once,
```
'0' disables recursion right? Why once? This might be difference in terminology. Thats why giving a quick example is better. Could you add this example?:

 Consider a simple simple recursive proto 'message Person { string name = 1; Person bff = 2}

> What would be spark schema when recursion 0, 1, and 2? I think : 

  - 0: struct<name: string, bff: null>
  - 1: struct<name string, bff: <name: string, bff: null>>
  - 2: struct<name string, bff: <name: string, bff: struct<name: string, bff: null>

SandishKumarHN replied:
@rangadi Thank you for your suggestion. I have implemented it by adding a comment and a unit test to make the example more clear to users.

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +693,435 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
+    }
+  }
+
+  test("Fail for recursion field with different field names without circularReferenceDepth") {
```
Fix `circularReferenceDepth` in the name.

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +693,435 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
+    }
+  }
+
+  test("recursion field with different field names with circularReferenceDepth") {
```
Fix the name.

rangadi replied:
What is this testing? As we discussed, field name does not matter.

rangadi commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:
```diff
@@ -693,4 +693,435 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
+    eventFromSparkSchema.getDescriptorForType.getFields.asScala.map(f => {
+      assert(expectedFields.contains(f.getName))
+    })
+  }
```
Could you add test that clearly shows the expected schema similar to my comment here: https://github.com/apache/spark/pull/38922/files#r1051292604

It is not easy to seem from these test what schema does 0 or 2 results in.

rangadi commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -1016,7 +1016,7 @@
   "RECURSIVE_PROTOBUF_SCHEMA" : {
     "message" : [
-      "Found recursive reference in Protobuf schema, which can not be processed by Spark: <fieldDescriptor>"
+      "Found recursive reference in Protobuf schema, which can not be processed by Spark by default: <fieldDescriptor>. try setting the option `recursive.fields.max.depth` as 0 or 1 or 2. Going beyond 3 levels of recursion is not allowed."
```
Why is 3 or above not alloweded? Seems pretty low. If a customer wants to set the level, they will be conscious. I think it should be at least high single digits to cover most cases. How about 10?

SandishKumarHN replied:
@rangadi agree.

rangadi reviewed (commented):
Overall LGTM. Made a couple of suggestions to clarify how schema looks like with recursive limit. Both in comment and unit test.

SandishKumarHN committed:
recursive depth to 10, unite tests, nit, comments

rangadi reviewed (approved):
This looks great. Thanks.

rangadi commented:
Asking @HeartSaVioR to take a quick look to approve. 
@cloud-fan take a look at the updated PR description for example of how spark schema would look like with the different setting for the config.

cloud-fan commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufOptions.scala:
```diff
@@ -38,6 +38,14 @@ private[sql] class ProtobufOptions(
+  // specified, the default value is -1; recursive fields are not permitted. If a protobuf
+  // record has more depth than the allowed value for recursive fields, it will be truncated
+  // and some fields may be discarded.
+  val recursiveFieldMaxDepth: Int = parameters.getOrElse("recursive.fields.max.depth", "-1").toInt
```
The option name may need a bit more discussion. Usually data source options do not have long names, and don't contains dot. See `JSONOptions`. How about `maxRecursiveFieldDepth`?

rangadi replied:
@cloud-fan this is in line with options for Kafka source. e.g. 'kafka.' prefix allows setting Kafka clientconfigs.

In addition, we will be passing more options. E.g. for schema registry auth configs. They will have a prefix like 'confluent.schemaregistry.[actual registry client conf]'

cloud-fan commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -40,19 +40,27 @@ object SchemaConverters {
-  def toSqlTypeHelper(descriptor: Descriptor): SchemaType = ScalaReflectionLock.synchronized {
+  def toSqlTypeHelper(
+      descriptor: Descriptor,
+      protobufOptions: ProtobufOptions): SchemaType = ScalaReflectionLock.synchronized {
```
not related to this PR, but why would we lock `ScalaReflectionLock` here?

cloud-fan replied:
cc @gengliangwang

rangadi replied:
Yeah, I just noticed. Not sure if if we need. 
@SandishKumarHN could we remove this in a follow up?

cloud-fan commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -40,19 +40,27 @@ object SchemaConverters {
   def structFieldFor(
       fd: FieldDescriptor,
-      existingRecordNames: Set[String]): Option[StructField] = {
+      existingRecordNames: Map[String, Int],
```
can we add comments to explain what map key and value means here?

rangadi replied:
+1

SandishKumarHN replied:
@cloud-fan added a comment.

cloud-fan commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -92,17 +108,35 @@ object SchemaConverters {
+        // 0: struct<name: string, bff: null>
+        // 1: struct<name string, bff: <name: string, bff: null>>
+        // 2: struct<name string, bff: <name: string, bff: struct<name: string, bff: null>>> ...
+        val recordName = fd.getMessageType.getFullName
```
```suggestion
        val recordName = fd.getFullName
```
are they same? The previous code uses `fd.getFullName`

rangadi replied:
Good catch. I think the previous code was incorrect. We need to verify if a same Protobuf type was seen before in this DFS traversal. 
@SandishKumarHN what was the unit test that verified recursion?

SandishKumarHN replied:
@cloud-fan fd.getFullName gives a fully qualified name along with a field name, we needed the fully qualified type name. we made this decision above. 

here is the difference.
```
println(s"${fd.getFullName} : ${fd.getMessageType.getFullName}")

org.apache.spark.sql.protobuf.protos.Employee.ic : org.apache.spark.sql.protobuf.protos.IC
org.apache.spark.sql.protobuf.protos.IC.icManager : org.apache.spark.sql.protobuf.protos.Employee
org.apache.spark.sql.protobuf.protos.Employee.ic : org.apache.spark.sql.protobuf.protos.IC
org.apache.spark.sql.protobuf.protos.IC.icManager : org.apache.spark.sql.protobuf.protos.Employee
org.apache.spark.sql.protobuf.protos.Employee.em : org.apache.spark.sql.protobuf.protos.EM
```
@rangadi previous code ```fd.getFullName``` fully qualified name along with a field name works to find out recursion. so before we just use to throw errors on any recursion field.

cloud-fan reviewed (approved):
looks good to me except for some minor comments

rangadi commented:
jenkins merge

SandishKumarHN committed:
review changes, comment for existingRecordNames

cloud-fan commented:
thanks, merging to master!

rangadi commented:
See #40011 for a follow up tweak for this config. '0' is not supported. Fixes how the limit is applied.
